Automattic/woocommerce-payments
# Fix race condition when processing webhooks

When the intent success webhook is received while processing the original request, it may result in duplicate `payment_complete` calls due to a race condition. This causes `woocommerce_order_status_processing` hook to fire twice resulting in duplicated emails, status change notes, and stock reductions.
Status checks introduced in #3287 alleviate the problem but race conditions still happen in the following scenario

1. Initial order processing starts. `process_payment()`->`process_payment_for_order`->`attach_intent_info_to_order`->`order->payment_complete`.
2. In parallel, `payment_intent.succeeded` webhook is received and we call `process_webhook_payment_intent_succeeded` -> `order->payment_complete`
3. Before we could update the order status to `processing` within the first `payment_complete` call, we pass through [this](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/admin/class-wc-rest-payments-webhook-controller.php#L304) status check introduced in #3287


Fixes #3345 

#### Changes proposed in this Pull Request

This change proposes mutexing `payment_complete` calls. A lock is placed before calling `payment_complete` in both above instances, which stops further processing if another attempt is made before unlocking or within five minutes of the initial call. Same approach is being used in [Stripe Gateway](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/trunk/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1505) to get around similar concurrency issues









#### Testing instructions





1. Listen to webhooks. `npm run listen` in your local server.
2. Enable UPE
3. To simulate the race condition, put a conditional `sleep` in woocommerce core `payment_complete` function. So that `payment_complete` invoked by webhook starts but doesn't change the status until the other request completes.
<details>
<summary> This is what I used </summary>

```
	public function payment_complete( $transaction_id = '' ) {
		if ( ! empty( $transaction_id ) ) {
			sleep(10);
		}
		if ( ! $this->get_id() ) { // Order must exist.
			return false;
		}

		try {
                .........
```
</details>

4. Notice that duplicated status changes in `develop` branch and no only one status change in this PR's branch

#### Additional Context

Previous attempt: https://github.com/Automattic/woocommerce-payments/pull/3399

-------------------


**Post merge**
